<div align="center">

![data](static/data.png)

# Spatial profiling of epigenomics via DBIT-seq

</div> 

Deterministic Barcoding in Tissue for spatial omics sequencing (DBiT-seq) uncovers spatial biology by combining microfluidics and next- generation sequencing (NGS). Applications of the platform have been published in [Nature](https://www.nature.com/articles/s41586-022-05094-1), [Nature Biotech](https://www.nature.com/articles/s41587-023-01676-0), [Science](https://www.science.org/doi/10.1126/science.abg7216), and [Cell](https://www.cell.com/cell/fulltext/S0092-8674(20)31390-8?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867420313908%3Fshowall%3Dtrue).

This repository contains the  scripts necessary to analyze data from ATAC and CUT&Tag DBIT-seq experiments performed with the barcoding scheme from [Zhang et al. 2023](https://www.nature.com/articles/s41586-023-05795-1#MOESM1) (barcodes on read2).

<div align="center">
  
[AtlasXomics](https://www.atlasxomics.com) • [Docs](https://docs.atlasxomics.com) • [Discord](https://discord.com/channels/1004748539827597413/1004748540624511008)

</div> 

# Workflow

**preprocessing** contains a single-cell epigenomic analysis pipeline built using [snakemake](https://bitbucket.org/snakemake/snakemake/wiki/Home); raw fastq files are uploaded directly to this directory for processing.  The pipeline is composed following steps,

- reads are filtered on the correctness of ligation linker sequences;
- alignment and preprocessing is performed with [Chromap](https://www.nature.com/articles/s41467-021-26865-w);
- BED output files are converted into fragment.tsv.gz files for downstream analysis.

> We recommend running the pipeline in a Linux environment with at least 64GB RAM, an 8-core Intel or AMD processor, and 1TB of free disk space; here, we provide instructions for running in an AWS EC2 instance.

**analysis** contains R scripts and vignettes for further analysis of epigenomic DBIT-seq experiments.  The scripts use the [ArchR](https://www.nature.com/articles/s41588-021-00790-6) package for dimensionality reduction, clustering, peak calling, and differential accessibility analysis and [Seurat](https://www.cell.com/cell/fulltext/S0092-8674(21)00583-3?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867421005833%3Fshowall%3Dtrue) for incorporation of spatial data.

> The **analysis** scripts require the 'spatial' folder generated by the [AtlasXBrowser](https://github.com/atlasxomics/AtlasXbrowser) package; see AtlasXomics [docs](https://docs.atlasxomics.com) for more information.

# Citations

Deng, Y., Bartosovic, M., Ma, S. et al. Spatial profiling of chromatin accessibility in mouse and human tissues. Nature 609, 375–383 (2022). https://doi.org/10.1038/s41586-022-05094-1

Liu, Y., DiStasio, M., Su, G. et al. High-plex protein and whole transcriptome co-mapping at cellular resolution with spatial CITE-seq. Nat Biotechnol (2023). https://doi.org/10.1038/s41587-023-01676-0

Yanxiang Deng et al. Spatial-CUT&Tag: Spatially resolved chromatin modification profiling at the cellular level.Science375,681-686(2022). DOI:10.1126/science.abg7216

Liu et al. High-Spatial-Resolution Multi-Omics Sequencing via Deterministic Barcoding in Tissue Cell. 183, 1665–1681 (2020). https://doi.org/10.1016/j.cell.2020.10.026

Zhang, H., Song, L., Wang, X. et al. Fast alignment and preprocessing of chromatin profiles with Chromap. Nat Commun 12, 6566 (2021). https://doi.org/10.1038/s41467-021-26865-w

Granja, J.M., Corces, M.R., Pierce, S.E. et al. ArchR is a scalable software package for integrative single-cell chromatin accessibility analysis. Nat Genet 53, 403–411 (2021). https://doi.org/10.1038/s41588-021-00790-6

Hao and Hao et al. Integrated analysis of multimodal single-cell data. Cell. 2021 Jun 24;184(13):3573-3587.e29. doi: 10.1016/j.cell.2021.04.048.
